File: app/src/main/java/cn/martinkay/wechatroaming/utils/HostInfo.java
package cn.martinkay.wechatroaming.utils;

import android.app.Application;

import androidx.annotation.NonNull;

/**
 * Helper class for getting host information. Keep it as simple as possible.
 */
public class HostInfo {

    public static final String PACKAGE_NAME_QQ = "com.tencent.mobileqq";
    public static final String PACKAGE_NAME_SELF = "cn.martinkay.wechatroaming";

    private HostInfo() {
        throw new AssertionError("No instance for you!");
    }

    @NonNull
    public static Application getApplication() {
        return cn.martinkay.wechatroaming.utils.host.HostInfo.getHostInfo().getApplication();
    }

    @NonNull
    public static String getPackageName() {
        return cn.martinkay.wechatroaming.utils.host.HostInfo.getHostInfo().getPackageName();
    }

    @NonNull
    public static String getAppName() {
        return cn.martinkay.wechatroaming.utils.host.HostInfo.getHostInfo().getHostName();
    }

    @NonNull
    public static String getVersionName() {
        return cn.martinkay.wechatroaming.utils.host.HostInfo.getHostInfo().getVersionName();
    }

    public static int getVersionCode32() {
        return cn.martinkay.wechatroaming.utils.host.HostInfo.getHostInfo().getVersionCode32();
    }

    public static int getVersionCode() {
        return getVersionCode32();
    }

    public static long getLongVersionCode() {
        return cn.martinkay.wechatroaming.utils.host.HostInfo.getHostInfo().getVersionCode();
    }

    public static boolean isInModuleProcess() {
        return cn.martinkay.wechatroaming.utils.host.HostInfo.isInModuleProcess();
    }

    public static boolean isInHostProcess() {
        return !isInModuleProcess();
    }

    public static boolean isAndroidxFileProviderAvailable() {
        return cn.martinkay.wechatroaming.utils.host.HostInfo.isAndroidxFileProviderAvailable();
    }

    public static boolean requireMinWechatVersion(long versionCode) {
        return getLongVersionCode() >= versionCode;
    }
}

File: app/src/main/java/cn/martinkay/wechatroaming/utils/HostInfo.kt
@file:JvmName("HostInfo")

package cn.martinkay.wechatroaming.utils

import android.app.Application
import android.content.ComponentName
import android.content.Context
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.util.Log
import androidx.core.content.pm.PackageInfoCompat

const val PACKAGE_NAME_SELF = "cn.martinkay.wechatroaming"

fun init(applicationContext: Context): HostInfoImpl {
    val packageInfo = getHostInfo(applicationContext)
    val packageName = applicationContext.packageName
    return HostInfoImpl(
        applicationContext,
        packageName,
        applicationContext.applicationInfo.loadLabel(applicationContext.packageManager).toString(),
        PackageInfoCompat.getLongVersionCode(packageInfo),
        PackageInfoCompat.getLongVersionCode(packageInfo).toInt(),
        packageInfo.versionName,
        when (packageName) {
            PACKAGE_NAME_SELF -> HostSpecies.BlackSpider
            else -> HostSpecies.Unknown
        },
    )
}

private fun getHostInfo(context: Context): PackageInfo {
    try {
        return context.packageManager.getPackageInfo(
            context.packageName,
            PackageManager.GET_META_DATA
        )
    } catch (e: PackageManager.NameNotFoundException) {
        Log.e("Utils", "Can not get PackageInfo!", e)
        throw e
    }
}

//val isInModuleProcess: Boolean
//    get() = hostInfo.hostSpecies == HostSpecies.BlackSpider
//
//val isInHostProcess: Boolean get() = !isInModuleProcess

data class HostInfoImpl(
    val context: Context,
    val packageName: String,
    val hostName: String,
    val versionCode: Long,
    val versionCode32: Int,
    val versionName: String,
    val hostSpecies: HostSpecies
)

enum class HostSpecies {
    BlackSpider,
    Unknown
}
